### ISSUE TITLE
Python SDK not building the confusion matrix example_matrix correctly

### ISSUE THREAD

**Issue by ironcadiz** [comet-ml/issue-tracking#374]
I was trying to make the rendering of images on the confusion matrix feature work following part 4 of this tutorial [https://www.comet.ml/docs/python-sdk/Comet-Confusion-Matrix/](https://www.comet.ml/docs/python-sdk/Comet-Confusion-Matrix/). Adapting it to my project (pytorch and pytorch-lightning) and using the latest version of the comet_ml library  I ended up with the following class:

```python
from torchvision import transforms
from comet_ml import ConfusionMatrix
import numpy as np

class ConfusionMatrixGenerator():
    def __init__(self, experiment, dataset):
        self.experiment = experiment
        self.dataset = dataset
        self.transform = transforms.ToPILImage()
        self.confusion_matrix = ConfusionMatrix(
            labels=["some_label", "another_label"],
            index_to_example_function=self.index_to_example
        )

    def log_matrix(self, epoch, targets, preds):
        self.confusion_matrix.compute_matrix(targets, preds)
        self.experiment.log_confusion_matrix(
            matrix=self.confusion_matrix,
            title=f"Confusion Matrix, Epoch #{epoch}",
            file_name=f"confusion-matrix-{epoch}.json",
        )

    def index_to_example(self, index):
        image_array = np.array(self.transform(self.dataset[index]['image']))
        image_name = f"confusion_matrix_{index}.png"
        results = self.experiment.log_image(
            image_array, name=image_name
        )
        return {"sample": image_name, "assetId": results["imageId"]} 
```
This successfully uploads the images to comet (they appear on the graphics tab) but the confusion matrix only shows the last uploaded example  per matrix cell instead of all the corresponding uploaded images.
Only way i could fix it is by modifying the `ConfusionMatrix` class on  file `utils.pyL1120`:

```python
 if self.index_to_example_function is not None:
        examples = []
        for index in retval:
            example = self._index_to_example(index, **kwargs)
            examples.append(example)
        if self._example_matrix[x][y] is not None:
           self._example_matrix[x][y] += examples
        else:
           self._example_matrix[x][y] = examples
```
Fixes the issue of only keeping the last uploaded asset on the matrix.

**Comment by dsblank** [comment]
@ironcadiz Thanks for the detailed report! I'll take a look at that.

**Comment by dsblank** [comment]
Indeed, you have correctly identified a bug. Fixing now... thanks so much!

**Comment by dsblank** [comment]
Actually, looking at this deeper, it seems that the idea isn't that the examples accumulate in the cells, but they only appear in the confusion matrix that mixed them up. For example, early on in training "a" might be confused with "b", and you would want to see that "a" in the mixed up cell. But later, when it is correctly identified as an "a" then it moves to the correct cell.

What work flow or example are you working on where the examples would accumulate?

**Comment by ironcadiz** [comment]
Thanks for the fast response!
Yes, the behavior  you describe is what I expect. My problem is when logging a single  matrix instance. When I mention the last uploaded image, I mean the last that was uploaded for that confusion matrix. 
For example in this case 16 examples appear in the cells and can be seen in the graphics tab but the matrix only shows 1 image per cell (which is the last uploaded image that belongs to that cell).

(had to censor the images)
![image](https://user-images.githubusercontent.com/13740457/90566298-37fdec80-e176-11ea-9104-8b974ce3a4e3.png)


With the change I described above  now it shows all the images correctly:

![image](https://user-images.githubusercontent.com/13740457/90566049-d3429200-e175-11ea-86d9-4a3db48213ea.png)

**Comment by dsblank** [comment]
Thanks for helping me understand your flow. I think there may still be a disconnect from how you are using the API. How many times do you call `log_matrix(epoch, targets, preds)` per epoch? If just once, then everything I think is ok. If more than once, then I could see your assumption breaking down.

**Comment by ironcadiz** [comment]
Just once on the `validation_epoch_end` step of pytorch-lightning

**Comment by dsblank** [comment]
Ok, I think you are correct. Thanks for the details. Much appreciated! Should have a fix out shortly, and some other enhancements.

**Comment by Lothiraldan** [comment]
Hi @ironcadiz, we are working on a fix for your issue and want to make sure to have properly reproduced it. I have some questions that should help me reproduce the issue:
* At the time you created the issue, you were using https://pypi.org/project/comet-ml/3.1.16/, right? And which version of Python are you using?
* The only parameters passed to the ConfusionMatrix in the non-redacted code are the same? Only labels and index_to_example_function?
* The datasets and targets and preds always have the same size (which seems to be 16 according to the shared screenshots), right?
* You are not using distributed training, neither multi-CPU or multi-GPU training, right? 

Would it be possible to see a redacted version of the Pytorch-Lightning `validation_epoch_end` which is calling the `log_matrix` method?

**Comment by ironcadiz** [comment]
* Yes versión 3.1.16 and  Python 3.6.8
* Yes
* No, the 16 corresponds  only to the pytorch-lightning sanity check validation pass (2 forwards of batch size 8). After that the dataset grows larger (~1k I believe)
* Right, only 1 gpu.

Here is a redacted version as you asked.
```python
def validation_epoch_end(self, outputs):
        avg_loss = stack_batch_metric(outputs, 'batch_val_loss').mean()
        y_pred = stack_batch_metric(outputs, 'batch_y_pred').flatten()
        y_true = stack_batch_metric(outputs, 'batch_y_true').flatten()
        preds = stack_batch_metric(outputs, 'batch_preds').flatten(0, 1)
        val_acc = accuracy_score(y_true.detach().cpu(), y_pred.detach().cpu())

        one_hot = torch.nn.functional.one_hot(y_true)
         if self.cm_generator is None:
            self.cm_generator = ConfusionMatrixGenerator(self.logger.experiment, self.dev_dataset)

        self.cm_generator.log_matrix(
            self.current_epoch,
            one_hot.cpu().detach(),
            preds.cpu().detach()
         )

        return {
            "val_loss": avg_loss,
            "log": {
                "val_loss": avg_loss,
                "val_acc": val_acc,
            }
        }

```

**Comment by Lothiraldan** [comment]
@ironcadiz I just released the Python SDK Version 3.2.1 that contains a fix for your issue, could you give it a try?
